src/Juvix/Core/Common/NameSymbol.hs

```diff
@@ -0,0 +1,39 @@
+module Juvix.Core.Common.NameSymbol where
+
+import qualified Data.List.NonEmpty as NonEmpty
+import qualified Data.Text as Text
+import Juvix.Library
+
+type T = NonEmpty Symbol
+
+toSymbol :: T -> Symbol
+toSymbol =
+  intern . foldr (\x acc -> unintern x <> "." <> acc) mempty
+
+fromSymbol :: Symbol -> T
+fromSymbol =
+  NonEmpty.fromList . fmap internText . Text.splitOn "." . textify
+
+subsetOf :: T -> T -> Bool
+subsetOf smaller larger =
+  case takeSubSetOfInternal smaller larger of
+    Just _ -> True
+    Nothing -> False
+
+takeSubSetOf :: T -> T -> Maybe T
+takeSubSetOf smaller larger =
+  case takeSubSetOfInternal smaller larger of
+    Just [] -> Nothing
+    Nothing -> Nothing
+    Just (x : xs) -> Just (x :| xs)
+
+takeSubSetOfInternal :: T -> T -> Maybe [Symbol]
```

Comment:
Ordering?

I guess it is a difference, but I'm taking the subset of the difference.

There is no assumed ordering it just eats the list from left to right until either the lists disagree or the smaller one ends

Reply:
That's not a subset, let's name the function something else. Can you clarify what the desired behaviour is? This will behave differently depending on the ordering of the lists, which seems strange to me given where this is used?

Reply:
I can rename it `takePrefixOf`

> This will behave differently depending on the ordering of the lists, which seems strange to me given where this is used?

The main usage throughout context 

```haskell
    Nothing ->
      case NameSymbol.takeSubSetOf currentName (s :| ymbol) of
        Just subPath ->
          updateCurr t <$> recurse f subPath currentNameSpace
        Nothing ->
          -- we do one more match as we have to make sure
          -- if we are adding foo we add it to the right place
```
is as above, checking if the currentName is a subpath of the above.  If the ordering were changed, then I can't determine that I need to lookup from current context.

Reply:
I see; `takePrefixOf` is better, yes